### ISSUE TITLE
torch.distributed.broadcast should default to current stream

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#19267]
## 🐛 Bug

This is the `torch.distributed` portion of #19219.

Repeated calls to `torch.distributed.broadcast` (and other distributed ops) with the NCCL backend will eat up all available GPU memory. This leads to repeated CUDA synchronizations and bad performance.

The `torch.distributed` ops should **default to the current stream** like all other PyTorch ops. DistributedDataParallel, which benefits from broadcasting in a background stream, should explicitly use a background stream (currently `ncclStream`) to keep good performance.

## To Reproduce

Here is a small repro. It should use a small amount of memory, but uses all available GPU memory (watch `nvidia-smi`):

```python
import torch
import torch.distributed as dist

def worker(rank):
    for itr in range(1000):
        x = torch.randn(int(25 * 1024 * 1024), device='cuda')  # 100 MiB
        dist.broadcast(x, src=1, async_op=False)
        del x

def main(rank, init_method, world_size):
    torch.cuda.set_device(rank)
    dist.init_process_group("nccl", init_method, rank=rank, world_size=world_size)
    worker(rank)

if __name__ == '__main__':
    init_method = 'tcp://127.0.0.1:23123'
    world_size = 2
    torch.multiprocessing.spawn(main, (init_method, world_size), nprocs=world_size)
```

## Additional context

- This wasn't an issue prior to #18465 because the distributed ops were missing calls to `record_stream`. (However, the ops behaved incorrectly and could overwrite memory that was in-use)
- Between #18465 and #19222 this led to out-of-memory errors
- After #19222 was landed, this generally does not lead to out-of-memory errors (due to synchronization, free, and retry on OOM) but still uses up all available memory

@ailzhang @teng-li @pietern @mrshenli

**Comment by pietern** [comment]
Thanks for doing the investigation and fixing this in the OOM case.

I walked through the relevant code paths and can't figure out why the existing implementation doesn't work. In your example, the allocated tensor will be free'd when you call `del` (or goes out of scope, etc). This means a call to `free(void *ptr)`, which means a call to `insert_events` because `stream_uses` is non-empty (per the current implementation of collectives in the NCCL backend). It then records an event on the NCCL stream (which will complete immediately because it is idle). Then when allocating the new tensor, it calls into `malloc`, which calls `process_events`, which will find the recently recorded event having completed, and making the block ready for allocation again.

Somewhere in this reasoning there must be a flaw, because clearly this is not what is happening. Perhaps the event that gets recorded on the call to `free` doesn't complete immediately, or there is some head of line blocking going on making the function return early? Head of line blocking could make this pile up, but that shouldn't be the case in your example. On the third call to `malloc` the first call to `free` will have its event completed for sure.

I'd like to understand this a bit better.

**Comment by colesbury** [comment]
> It then records an event on the NCCL stream (which will complete immediately because it is idle)

The event doesn't complete immediately. The NCCL broadcast has been "launched" asynchronously, but it hasn't completed yet. (The `randn` is launched asynchronously too).

The CPU can get through the for-loop many times while the `randn` and `ncclBcast` kernels are waiting to be executed in some internal CUDA queue.

Here's a modified version of the code that demonstrates the delay between launch and completion. It outputs the "delay" in loop iterations between when a `randn` call is launched and when it's finished. A delay of 0 means it's done by the next loop iteration. The output looks like:

```python
[0, 10, 20, 25, 35, 44, 54, 58, 68, 77, 87, 95, 104, 113, 117, 116, 115, 114, ...]
```

i.e. the first `randn` launch is complete by next time through the loop, but the second `randn` launch isn't done until 10 iterations later.  The delay is capped at ~117 because after that all the memory of the GPU is allocated (100MB*117 = ~11.7GB), which forces a full synchronization and cudaMalloc retry.

Code is here:

```python
import torch
import torch.distributed as dist

def worker(rank):
    event = None
    torch.cuda.synchronize()
    events = [torch.cuda.Event() for _ in range(1000)]
    delay = [0 for _ in range(1000)]
    offset = 0
    for itr in range(1000):
        while offset < itr and events[offset].query():
            delay[offset] = itr - offset - 1
            offset += 1
        x = torch.randn(int(25 * 1024 * 1024), device='cuda')  # 100 MiB
        events[itr].record()
        dist.broadcast(x, src=1, async_op=False)
        del x
    print(delay[:offset])  # only print up to the kernels that have finished

def main(rank, init_method, world_size):
    torch.cuda.set_device(rank)
    dist.init_process_group("nccl", init_method, rank=rank, world_size=world_size)
    worker(rank)

if __name__ == '__main__':
    init_method = 'tcp://127.0.0.1:23123'
    world_size = 2
    torch.multiprocessing.spawn(main, (init_method, world_size), nprocs=world_size)
```

**Comment by pietern** [comment]
Thanks, this makes things clear. The repro illustrates the problem perfectly. I was thinking of the underlying implementation, where it kicks off a collective, and then calls `wait()`. What I forgot here is that `wait()` for NCCL is also asynchronous (it only executes `cudaStreamWaitEvent` for completion of the NCCL kernels).

I also understand your solution to simply launch everything on the default stream. This allows the allocator to reuse the same block over and over again, since everything is queued on the same stream anyway.

The reason the NCCL process group uses a separate stream is not only to allow for overlapping, but also to enforce sequencing. Separate streams is not a hard requirement here, it just means we get the sequencing for free. We could turn this around and record events on whatever stream the user has set and sprinkle `cudaStreamWaitEvent` to force sequencing of the NCCL kernels that way. We could do the same for the Gloo backend, where we use a separate set of streams to perform copies to/from pinned memory (this happens on a separate thread).

That said, there is another issue that's relevant here. The ProcessGroup API at the C++ level returns a `ProcessGroup::Work` object for every operation. The work is done in the background and the caller calls `Work::wait` to synchronize on completion of the work. This is also where errors are bubbled up, if they happened (e.g. timeout or other I/O errors). This is how we avoid blocking the main thread regardless of backend. For the Gloo backend, the call to `wait` means a check if the operation resulted in an error or not, and optionally (when working with CUDA tensors), a stream synchronization on the results being written back in place. This is why the repro doesn't repro when using the Gloo backend. The call to `wait` for the NCCL backend is currently only a stream synchronization with the events that record the end of the respective operation, which allows for this "unlimited queuing" of CUDA operations, but also prevents us from implementing #17882.

Summarizing:

* Use of multiple streams in combination with "unbounded queuing" causes this issue
* We'd like to be able to detect I/O errors when executing NCCL operations
* Detecting errors for NCCL operations means synchronizing on completion
* Synchronizing on completion means preventing unbounded queuing

If we go down this path, it no longer matters whether we run these operations on the user stream on a new stream, because all work will have completed by the time `wait` returns.

What do you think?

**Comment by leonardo0lyj** [comment]
Any luck to solve this issue -- "repeated `torch.distributed.broadcast` eats up all available GPU memory"?
How to force "repeated `torch.distributed.broadcast`" to reuse the same memory block? (Instead of blowing up the memory and then free everything.) 

Really appreciated!
